akka-remote/src/main/scala/akka/remote/artery/OutboundEnvelope.scala

```diff
@@ -52,16 +56,30 @@ private[akka] final class ReusableOutboundEnvelope extends OutboundEnvelope {
   private var _recipient: OptionVal[RemoteActorRef] = OptionVal.None
   private var _message: AnyRef = null
   private var _sender: OptionVal[ActorRef] = OptionVal.None
+  private val _metadata: MetadataMap[AnyRef] = MetadataMap()
 
   override def recipient: OptionVal[RemoteActorRef] = _recipient
   override def message: AnyRef = _message
   override def sender: OptionVal[ActorRef] = _sender
+  override def metadata: MetadataMap[AnyRef] = _metadata
 
   override def withMessage(message: AnyRef): OutboundEnvelope = {
     _message = message
     this
   }
 
+  /**
+   * IMPORTANT: The Byte keys must be greater than 0 and less than 32.
+   * These are refered to as metadata-slot keys. They should be statically assigned
+   * on a tool-by-tool basis.
+   *
+   * The keys 0–7 are reserved for Akka internal purposes and future extensions.
+   */
+  def setMetadata(id: Byte, metadata: AnyRef): OutboundEnvelope = {
+    _metadata.set(id.toInt, metadata)
+    this
+  }
+
   def copy(): OutboundEnvelope =
     (new ReusableOutboundEnvelope).init(_recipient, _message, _sender)
 
```

Comment:
GitHub doesn't allow me to comment on existing code, but I'm missing a call to `_metadata.clear()` in the `clear()` method just below here.